generate_charts.py

```diff
@@ -0,0 +1,206 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+Modifications are licensed under
+
+Apache License, Version 2.0
+(see ./LICENSE for details)
+
+"""
+
+import pathlib
+
+import diskcache
+import pandas as pd
+import plotly.express as px
+import plotly.graph_objs as go
+from dash import DiskcacheManager
+
+cache = diskcache.Cache("./cache")
+background_callback_manager = DiskcacheManager(cache)
+
+from app_configs import RESOURCE_NAMES, SCENARIOS
+from src.model_data import JobShopData
+
+BASE_PATH = pathlib.Path(__file__).parent.resolve()
+DATA_PATH = BASE_PATH.joinpath("input").resolve()
+
+model_data = JobShopData()
+
+
+def get_minimum_task_times(job_shop_data: JobShopData) -> pd.DataFrame:
+    """Takes a JobShopData object, gets the minimum time each
+    task can be completed by, and generates the Jobs to Be Scheduled
+    Gantt chart.
+
+    Args:
+        job_shop_data (JobShopData): The data for the job shop scheduling problem.
+
+    Returns:
+        pd.DataFrame: A DataFrame of the jobs to be scheduled including Task, Start, Finish,
+        Resource, and delta.
+    """
+    task_data = []
+    for tasks in job_shop_data.job_tasks.values():
+        start_time = 0
+        for task in tasks:
+            end_time = start_time + task.duration
+            task_data.append(
+                {
+                    "Job": task.job,
+                    "Start": start_time,
+                    "Finish": end_time,
+                    "Resource": task.resource,
+                }
+            )
+            start_time = end_time
+    df = pd.DataFrame(task_data)
+    df["delta"] = df.Finish - df.Start
+    df["Job"] = df["Job"].astype(str)
+    df["Resource"] = df["Resource"].astype(str)
+    return df
+
+
+def get_empty_figure(message: str) -> go.Figure:
+    """Generates an empty chart figure message.
+    This is used to replace the chart object
+    when no chart is available.
+
+    Args:
+        message (str): The message to display in the center of the chart.
+
+    Returns:
+        go.Figure: A Plotly figure object containing the message.
+    """
+    fig = go.Figure()
+    fig.update_layout(
+        xaxis={"visible": False},
+        yaxis={"visible": False},
+        annotations=[
+            {
+                "text": message,
+                "xref": "paper",
+                "yref": "paper",
+                "showarrow": False,
+                "font": {"size": 28},
+            }
+        ],
+    )
+    fig.update_layout(
+        margin=dict(l=20, r=20, t=10, b=10),
+    )
+    return fig
+
+
+def generate_gantt_chart(
+    scenario: str = None, df: pd.DataFrame = None, y_axis: str = "Job", color: str = "Resource"
+) -> go.Figure:
```

Comment:
This could be updated to take the `model_data` instead of `scenario` which would leave both the `src.model_data.JobShopData` import and the `SCENARIOS`, `RESOURCE_NAMES` and, especially, `DATA_PATH` variables in `app.py` instead of declaring/importing them into this file as well.

Reply:
updated: https://github.com/dwave-examples/job-shop-scheduling-cqm/pull/17/commits/5e41bd2e6de63f470bc2ac073b985861d81f7698